docs/architecture/adr-005-subscription.md

```diff
@@ -0,0 +1,225 @@
+# ADR 5: Subscription
+
+## Changelog
+
+
+## Context
+
+Cosmos-SDK should allow developers to choose from a variety of payment models for their application. One of the most popular payment models for Internet applications, so it is natural to provide this payment model as a module for sdk users.
+
+For fuller context around this issue: see [\#4642](https://github.com/cosmos/cosmos-sdk/issues/4642)
+
+## Decision
+
+Create Subscription module in SDK so that users can subscribe to various subscription services. Subscription collectors will send messages to collect subscription payments from users who have due subscriptions. Subscriptions are paid out directly from a user account. If the user does not have enough funds to pay for the subscription is inactivated. Users can define maximum limits on how many periods the subscription is valid for; if this limit elapses before a subscription renewal, the subscription is invalidated.
+
+## Status
+
+#### Proposed
+
+## Implementation Changes
+
+Introduce the following messages to `x/subscription` module:
+
+```go
+// CreateSubscriptionMsg allows Service Provider to create new subscription service.
+type CreateSubscriptionMsg struct {
+    Name      string         // unique, human-readable name for subscription service
+    Amount    sdk.Coins      // amount to be collected for each subscription period
+    Period    time.time      // duration of subscription period
```

Comment:
this should be `time.Duration`. 

Also, most of the subscriptions allow for a set of subscription periods (1mo, 3mo, 6mo, 1y) with their corresponding prices (i.e registering an `Amount` to a `Period`). 

I think it'd be better to extend this to be an array. Otherwise, the user will have to create multiple messages for each of these periods.

Reply:
Hmm yea, so the current FIFO queue currently works with the assumption that there is a single period and the Collector never gets to change (at least never shorten) this period.

Probably too restrictive. Would like to hear more about the tree idea, a bit fuzzy on what you mean.